Semantic search added

ndrean opened the pull request:
@luchoturtle 

First trial. Still amazed to see what you can do.

It works, probably needs a serious read. I close the other PR.

The embedding model is started via a GenServer. Maybe we can or should include this in your Models.ex.

1) Reset your the DB, and run the server. Logs should display "New Index".

2) Trials:

- load 2 different images, say a house and then a dog. Then record a short audio, say a house. Check the result.
- stop & start the server. Logs should output "Existing Index". Record an audio for a dog. Check the result.

> Note:  if you erase the Index file without resetting the DB, there will be obviously a problem.

I experience that the semantic search is not perfect, especially if the audio transcription fails. It also depends a lot on the words you use. If you ask for a picture of a cat, you may have a dog or a house..... This is where a threshold is needed. TBC!

LuchoTurtle commented:
I've merged #43 to `main`  :)

ndrean commented:
More remarks:

- the HTML can obviously be improved
- Let me know if the last part is "pedantic". I felt that the documentation of HNSWLib could be improved so I adjoined something here, but it can be censored.
- I tried several times to find a threshold by looking at the "distance" or angle. I output the distance on each search, so you can observe it as well. My best guess is not use the result when the distance is above 0.12-0.15, but we need more fine-grained explains.

ndrean commented:
I added "save Index file to Postgres'" to overcome the Fly volume problem as it is crucial to keep the "images" table and the HNSWLIb Index file aligned

ndrean commented:
At this point, I believe the code works, once again except the tests and Readme. I didn't write much since I don't know if you accept the code as such. Except for the models, it relies on Postgres for the images and Index so it should be deployable. I am very very curious to see this online!!

LuchoTurtle commented:
Hey @ndrean ! 

Before anything, thank you for the changes! Let me go over some of the points you've raised.

---

> about unique uploads. Nothing prevents a user from uploading several times the same image. 

The images are being uploaded to https://github.com/dwyl. The name of each file stems from the contents of the image, which is then hashed. So if we upload two images to https://github.com/dwyl that are the same, we're just replacing one with the other :)

> the UX: I should remove the previous found picture & transcription once the semantic search runs. We should probably do the same for the caption.

The UX can be taken care of on another PR. It's easier to review smaller PRs then big ones :D Plus, I have some ideas on how to tackle this later.

> second UX point: the caption is not editable, nor is the audio transcript. This could be a refinement that you may want so the user can modify them in case of an error. 

This is a neat idea. But this is just a small demo/project to learn how to get some ML running on Elixir apps, it's not meant to be a production-ready app. Those details are cool but they're not the focus of this small project :D 

---

Related to saving the index file in Postgres, I believe you made the right call (since we're deploying this to `fly.io`. I'm aware that there's an extension to Postgres called https://github.com/pgvector/pgvector, which is basically a vector database (which can inclusively perform knn search). But I haven't found any links of people able to run this extension in `fly.io` machines.

Your strategy to load the index file seems to make sense. I honestly have no idea how much time it takes to save the file to the DB but it seems that `Ecto,Multi` does **not** lock the table (from the link you sent me). It seems that, if we want to have a pessimistic row-level locking procedure going on, we have to use the [`lock`](https://hexdocs.pm/ecto/Ecto.Query.html#lock/3) function instead.

However, do we actually *really want pessimistic locking*? I think if there's a race condition, we can either return the info to the user with feedback or let it resolve itself (https://stackoverflow.com/questions/54896663/what-is-a-real-world-example-where-you-would-choose-to-use-pessimistic-locking).  We can try both options and see how it fares though, for sure (at least `Logger` it, so we know in case it happens).

ndrean commented:
Thanks Lucho!

About the SHA:

> we're just replacing one with the other :)
ok, but:
- you still can create 2 indices for the same picture, and this is not good since there is some kind of clustering, so I don't know the consequences
- and is it worth going through all the process, making another call to S3 ect whilst you can simply add the SHA (its a one line computation in Elixir) as an indexed field into the Image table and look-up of it and make an early stop? This is what I implement, very little changes. I will even put a function in the Image.ex so that the 2 lines will become one with any apparent call to the DB.

The usage is pretty reasonable and short and stops the process quite early. I just add an indexed field to the Image table,  and add the SHA1 string as a field to the `%ImageInfo{}` struct, hat's pretty much it.

```elixir
with {:magic, {:ok, %{mime_type: mime}}} <-
                    {:magic, magic_check(path)},
                  file_binary <- File.read!(path),
                  sha1 <- App.Image.calc_sha1(file_binary),
                  {:sha_check, :ok} <- {:sha_check, App.Image.check_sha1(sha1)},
                  {:image_info, {mimetype, width, height, _variant}} <-.....

else
  {:sha_check, nil} -> push_event(socket, "toast", %{message: "Already present"})
```

with

```elixir
def calc_sha1(file_binary) do
    :crypto.hash(:sha, file_binary)
    |> Base.encode16()
  end

  def check_sha1(sha1) do
    App.Repo.get_by(App.Image, %{sha1: sha1})
    |> case do
      nil ->
        :ok

      _ ->
        nil
    end
  end
```


The Index file is not a vector, and seems small ? (10kB for 5 entries) and takes 4 ms. I believe this is reasonable?

> `pg_Vector`: But I haven't found any links of people able to run this extension in fly.io machines.
I guess you can use extensions, such as PostGIS. See [this](https://community.fly.io/t/postgres-with-postgis/836) thread. But it is clear as mud, sorry.

LuchoTurtle commented on lib/app/image.ex:
```diff
@@ -8,26 +8,53 @@ defmodule App.Image do
   def changeset(image, params \\ %{}) do
     image
-    |> Ecto.Changeset.cast(params, [:url, :description, :width, :height])
+    |> Ecto.Changeset.cast(params, [:url, :description, :width, :height, :idx, :sha1])
```
I'm looking over the code. Can you remind me why we need `idx` and `sha1`? I assume they're related to the indexing for the semantic search, but can you elaborate further so I understand it? 😅

LuchoTurtle replied:
Nvm, you've just answered it in https://github.com/dwyl/image-classifier/pull/47#issuecomment-1909115492.

ndrean replied:
Yes, `sha1` is the string to uniquely identify a file, and `idx` is the indice of an entry in the "indexes.bin" as returned by `HNSWLib`: this uniquely identifies a caption embedding <-> unique image.
Is this more clear?

LuchoTurtle replied:
Quite indeed. Thank you very much! :D

ndrean replied:
Just realized that we are doing a look-up on this "idx" column in the Image table, so you need to `create unique index` in the migration/schema for performance.

ndrean commented:
I looked at [optimistic locks](https://hexdocs.pm/ecto/Ecto.Changeset.html#optimistic_lock/3). It seemed pretty easy to use, so I added it, just a few changes, exactly as per the doc. The idea is that the Hnswlib Index can only be saved when it is not used by another user, in which case I ask for a retry.
Furthermore, I had to modify the Ecto.Multi failure treatment: within a transaction, you firstly update the image, then save the Idnex into the DB and handle the errors. 
In fact, I did not take into account the fact that 2 different users may upload the same image at almost the same time. In this case, the SHA won't work because the image save to the DB occurs after the caption is computed, which means a few seconds, which gives space for another user to upload the same image (thus with the same SHA) but the DB is not yet uplated with the SHA. I hope this makes sense.

ndrean commented:
The code before is good for a single user, but I circled a bit around with race conditions when openning 2 screens. 
The problem I tested is when you open 2 screens, and load the same image. 
I firstly saved the Image into the DB once you get the SHA, the URL and the caption and the HNWSLib index, then you have plenty of opportunities to enter several HNSWLib indexes for the same image.
Then I decided to separate the DB save of an image. Save a partial image once you get the SHA so another user get an early stop: no upload to the bucket, no caption, no embedding and no entry in the Index.
Then if by any chance he passes the SHA test, I introduced another SHA check before append the Index. This should definitively stop  multiple indexes for the same image.
Lastly, I lock the Index file saving into the db, since different users may want to save the Index file. This file should be the same for different users (on the same machine!).
My last push must solve this problem.

ndrean commented:
Maybe some useful notes:
1) which model to use? This may be the most important point, but to be honest, I simply used the ones I found in Sean Moriarty articles.
2) how to run a KNN search? This I can explain a bit more below.

As @LuchoTurtle already mentioned , there is the [pgvector extension](https://github.com/pgvector/pgvector). This was my first thought. Once you compute the embedding of an image caption , you update your image record and store the vector in the DB as a (pg)-vector. Then you query with the distance operator (`<->`). I already used in a project (to find existing GPS points when you navigate in a map in my LiveMap). It works well. Since we are only using Postgres, this appeared to me as a healthy solution: you can perform a CRUD on an image, the source of truth is the DB, so looked fine.

So why HNSWLib and not pgvector? I did not found a source  that explains how to implement the extension with Fly. And  @luchoturtle seemed to confirm this.
Since I do not have a Fly machine, I did not even try this.
Then I discovered Faiss and HNSWLib when reading [the article](https://dockyard.com/blog/2023/03/07/audio-speech-recognition-in-elixir-with-whisper-bumblebee?utm_source=elixir-merge) written by Sean Morairty. He used [ExFaiss](https://github.com/elixir-nx/ex_faiss) but this library is archived. They recommend HNSWLib. So I went for it.

Hnswlib relies on a physical Index file. 
To convince myself it works, I open 2 browsers and use an empty DB. Then I upload a picture in a browser, and another one in the other browser, and record an audio on say the first browser to look for the picture loaded in the second browser, and it finds it. Nice. So this works.
You can't prune an index in HNSWLib so if I load the same picture twice, it appears twice in the Index. I believe this is a problem simply because if you ask for say the 3 nearest vectors, then you will have this picture twice. Not good. 

These are the reasons for all this gymnastic.
I also had [a problem](https://the-eef.slack.com/archives/C01PU0RDJ8L/p1701952144024739) with the cosine similarity. As you see, I solved this by using unit vectors.

So if not Postgres-pgvector with Fly, if not Postgres-HNSWLib, why not SQlite and HNSWLib? I found [this but in Python](https://github.com/jiggy-ai/hnsqlite). So If you want Fly and Postgres, then I saw [Supabase](https://fly.io/docs/reference/supabase/)  that integrates [pgvector]((https://github.com/supabase/postgres/pull/472)).
Do you have experience with Supabase?

LuchoTurtle commented:
I always have pacience to read your awesome texts @ndrean :P 

Thanks for further explaining choosing HNSWLib over pgvector. It's really useful to include this in the README afterwards so people understand the reasoning. 

And thank you so much for the engineering and explaining the process of getting the Index files to not deadlock and to avoid conflicts. It's really interesting seeing how you went sorting this out under the constraints of `fly,io`. To be honest, didn't know that updating the SHA would entail this much effort to avoid racing conditions. But glad you got a way around it, it looks awesome ⭐ .

Regarding your cosine similarity original comment, can you post a screenshot? I don't have access to the Slack channel and it's easier for people to see the original question :). Can you expand on why unit vectors performed better? I thought using the appropriate metric (closing similarity, for example) when `knn`-ing depended on the metric that was used in the embedding model 🤔 .

I've tried Supabase before but just for a little demo (https://github.com/dwyl/supabase-flutter-demo). It was just to experiment the framework but I didn't delve THAT deep into it, since I only used it as a SPIKE experiment.

ndrean commented:
To get a felling of HNSWLib, I ran an IEX session:

```elixir
Mix.install([
{:bumblebee, "~> 0.4.2"},
{:exla, "~> 0.6.4"},
{:nx, "~> 0.6.4 "},
{:hnswlib, git: "https://github.com/elixir-nx/hnswlib", override: true}
])

defmodule Embedding do
  use GenServer
  @indexes "indexes.bin"

  def start_link(_) do
    GenServer.start_link(__MODULE__, {}, name: __MODULE__)
  end

  # upload or create a new index file
  def init(_) do
    space = :l2
    # space = :cosine

    {:ok, index} =
      case File.exists?(@indexes) do
        false ->
          HNSWLib.Index.new(_space = space, _dim = 384, _max_elements = 200)

        true ->
          HNSWLib.Index.load_index(space, 384, @indexes)
      end

    model_info = nil
    tokenizer = nil
    {:ok, {model_info, tokenizer, index}, {:continue, :load}}
  end

  def handle_continue(:load, {_, _, index}) do
    transformer = "sentence-transformers/paraphrase-MiniLM-L6-v2"

    {:ok, %{model: _model, params: _params} = model_info} =
      Bumblebee.load_model({:hf, transformer})

    {:ok, tokenizer} =
      Bumblebee.load_tokenizer({:hf, transformer})

    {:noreply, {model_info, tokenizer, index}}
  end

  def serve() do
    GenServer.call(__MODULE__, :serve)
  end

  def get_count do
    GenServer.call(__MODULE__, :get_count)
  end

  def get_index do
    GenServer.call(__MODULE__, :get_index)
  end

  def handle_call(:serve, _from, {model_info, tokenizer, index} = state) do
    serving = Bumblebee.Text.TextEmbedding.text_embedding(model_info, tokenizer)
    {:reply, {serving, index}, state}
  end

  def handle_call(:get_count, _, state) do
    {:ok, count} = HNSWLib.Index.get_current_count(state)
    {:reply, count, state}
  end

  def handle_call(:get_index, _, state) do
    {:reply, state, state}
  end
end
```

then continue and start the GenServer:

```elixir
{:ok, pid} = GenServer.start_link(Embedding, [])
{serving, index} = GenServer.call(pid, :serve)
```

Let's compute embeddings for "small" and "tall" and append it to the Index.

```elixir
%{embedding: data} = Nx.Serving.run(serving, "small")
HNSWLib.Index.add_items(index, data)

%{embedding: data2} = Nx.Serving.run(serving, "tall")
HNSWLib.Index.add_items(index, data)
```

Then let's search for the nearest neighbour for "high".

```elixir
%{embedding: data3} = Nx.Serving.run(serving, "high")
{:ok, labels, distances} = HNSWLib.Index.knn_query(index, data, k: 1)
idx = Nx.to_flat_list(labels[0])
{:ok, dt} = HNSWLib.Index.get_items(index, idx)
Nx.stack(Enum.map(dt, fn d -> Nx.from_binary(d, :f32) end))
```

Check your result. The label is the indice in the Index (not obivous, isn't it? 😀). The embedding should be the embedding copmputed for "small".

Do the same thing but endow the vector space with `space= :cosine`. You will see that the found embedding is not the same.

To solve this, "normalize" the vector:
```elixir
%{embedding: data} = Nx.Serving.run(serving, "small")
data = Nx.divide(data, Nx.LinAlg.norm(data))
HNSWLib.Index.add_items(index, data)
```

Then you should recover the embedding.

What's the point of doing this?
Well, if you append the Index file item by item, then you have the indice and can save it to the DB to uniquely identify an item. You can then look-up for an indice with the `knn_search`.

But, if you enter a batch of items, then you don't have the indice, but only the embedding vector. Now you saw that `knn_search` doesn't return a known embedding. But if you normalize it, then the `knn_search` returns a correct embedding.

That is the reason of the normalisation. I hope this helps to understand all this 🤯

ndrean commented:
I mainly refactored the "embedding" model so that it is now loaded in the general "App.Models" module.

I also discovered some edge cases. You will argue that tests should be written by now. The long running processes makes it a bit tricky and some race conditions appeared when I tried to hammer a bit the code with several sessions simulating multiple users. Some bugs appeared in these conditions and I addressed them in this new commit.

Note that when you use a `GenServer` with `handle_continue` to load a model, it happens asynchronously. Now it doesn't.
I believe there might be an opportunity to explore to potentially speed-up the first load (to be tested). This however will make testing more difficult because the "App.Models" would disappear for individual GenServers.

- when you start a fresh app, you load the models and create a new Index. However, it appeared that when you stop the app without entering any embedding in the Index, then there is a bug because I save a void Index in the DB so I have to erase it and recreate it.

- using a unique "tmp.wav" file to store the audio is wrong. So I give a unique name (with an `Ecto.UUID.generate()`) to each tmp.wav file assigns to a user <=> liveview session. 

- I had to assign a temporary random value to the "idx" field of each image when I first save the SHA. Recall that the "idx" field  is the order of insertion of the embedding of an image in the Index file. This "idx" is the returned value of the KNN search and uniquely identifies an image. I assigned this is because I set a unique constraint on the "idx". I did this with `:rand.uniform` and set a `bigint` as the Postgres type. Once the embedding is computed and inserted into the Index, the "idx" field is updated to the correct value. Note that it is easier /faster to run a look-up on such an integer than on a vector field such as an embedding, so I hope this makes sense.

- I have had an `Ecto.StaleEntryError` error. I couldn't repeat it on demand. I believe  - but not sure - that this is linked to the lock strategy when I save the Index into the database (needed to restart a session correctly, to have the Image table synced with the Index table). I added a `try/rescue` but never it never triggered again unfortunately.

I feel there is a refactoring opportunity on the general `handle_info` where I used `cond do`. I did not do it because this breaks the first captioning text.

Thss commit reflects these changes.

ndrean commented:
The  Python CLI utility  `LLM` to compare with the Bumblebee results.
[<img width="812" alt="Screenshot 2024-02-03 at 16 23 11" src="https://github.com/dwyl/image-classifier/assets/6793008/3511315e-916c-46a7-912c-c26c658e5787">](https://llm.datasette.io/en/stable/embeddings/cli.html)


Install `llm` and test some embeddings:

```bash
llm embed -m sentence-transformers/all-MiniLM-L6-v2 -c "small"
```

LuchoTurtle commented:
Thank you so much for the **very thorough** explanation, @ndrean ! Now I understand the topic much better! I'm sure I'll use your explanation in the `README` for everyone to read!

I noticed that the code may become a bit convoluted since you want to fix all the possible race conditions. I understand the reasoning but this project is simply a showcase of what `Elixir` can do, it's defo not supposed to be production-ready. Tho, at this point, it's a **very good** baseline for it :P 

If you feel like it's completed and good for review, ping me and I'll see to it :)

ndrean commented:
Thank you so much for your patience to read all this. 

If you tested HNSWLIB in IEX, then you have a feeling.

You are right, becomes complicated to read, unless you wrote it. This is a bad sign? 

And you are right, it is not expected to be production ready, but I hate to let a function return {:ok, }} | {:error, _}  and not tackle both cases.

Now there were plenty of dialyzer and credo warnings so I tried to refactor and slice the code. and added guards on types when necessary.
The `with` statement and the `|>` makes it easier to read I believe. 

The problem is to remember what does each function, and why all this luxury of tests before saving.
If a single user plays with the app, then no need for all this, 100% agree.
I just encounter quite a few errors when I stated to open several browsers, to see how I could hammer this poor little app.

The other point is that you need to save the Index file not only in the Filesystem but also in the DB, and this is to handle the Fly pruning.


Recall the aim is to have a prefect synchronisation between the Index file and the database. 

Every time I come back to this, I find refactoring opportunities, and I refrained from slicing the code even more!

I still need to add things in the Readme, to make it easier to follow the reasoning. But obviously, I need a new eye and more.

ndrean commented:
@LuchoTurtle 

You may a look at the code at this stage.
The Readme needs a lot of corrections, but most importantly is the code.

Tests I am starting to write:

1) Start with fresh:  mix ecto.drop && mix.create && mix.migrate && rm -rf priv/static/uploads

2) If you stop and restart, it should be ok

3) if you upload an image, then stop, then `rm -rf priv/static/uploads`, it should load the Index from the db

4) If you drop the db, (and keep the Index), the server can't run.

5) upload an image, and then `delete_all(App.Image)`. Open another browser and you should be redirected with the `on_mount` to a 404.

6) images can't be uploaded twice

7) Stop and erase say ".bumblebee/paraphrase-MiniLM-L6-v2", it should download this model.

I tried to be careful with the race conditions, and be able to rollback the db before an insertion in the Index file to keep the coherence (crucial).

The Genserver holds the Index in the state, so is centralised. This means this app can't be distributed as such.

LuchoTurtle commented:
Thank you @ndrean . I'll take a look at it once I can. I feel like I'll need some time to review this PR, it's quite big :P

ndrean commented:
Here to help! But run it, you will be surprised by the result.
I really thought how to make small changes, but then you are constantly changing the previous code as soon as you add something, or address an issue.

ndrean commented:
I forgot to pop a small stash. Besides this and  the big handle_info to refz tor (if you want), it's ok!

LuchoTurtle commented:
Just tried uploading an image and recording my voice to search it and it seems to be working! :D 

<img width="880" alt="image" src="https://github.com/dwyl/image-classifier/assets/17494745/c9c9a503-f97f-4160-b942-72b06141f415">

Though, I was getting an error with `hnswlib` when first trying to `mix phx.server`.

I was getting the following error:

```
Error happened while installing hnswlib from precompiled binary: "missing checksum.exs file".

Attempting to compile hnswlib from source...
/bin/sh: cmake: command not found
incomplete build of hnswlib found in '/Users/lucho/Documents/dwyl/image-classifier/_build/dev/lib/hnswlib/cmake_hnswlib', please delete that directory and retry
make: *** [/Users/lucho/Documents/dwyl/image-classifier/_build/dev/lib/hnswlib/priv/hnswlib_nif.so] Error 1
could not compile dependency :hnswlib, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile hnswlib --force", update it with "mix deps.update hnswlib" or clean it with "mix deps.clean hnswlib"
==> app
** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. Try running the
commands "gcc --version" and / or "make --version". If these programs
are not installed, you will be prompted to install them.
```

I had both `gcc` and `make` installed, so that wasn't the issue. 

I noticed that inside `mix.exs`, you had `{:hnswlib, git: "https://github.com/elixir-nx/hnswlib", override: true},`. I switched it to `{:hnswlib, "~> 0.1.4"},` and everything seems to be working.

Is there a reason you were using the git url?

ndrean commented:
> Is there a reason you were using the git url?

Yes: <https://github.com/elixir-nx/hnswlib/issues/16>

Try several images, otherwise it will always the same for sure as there is no cut-off distance. I mentioned this as the main residual potential problem.
I tried things, never convinced so I don't know. Maybe the distance > 0.2???? This is really tricky, isn't it?

LuchoTurtle commented:
Fair enough. I got it running :D 

However...

I'm trying to run `mix test` to see if I can do some basic testing. Initially, I couldn't because I was getting a `"Incoherence on table"` error. I noticed that this was because the GenServer was being executed and there was no test data for the GenServer to occur.

```elixir
        App.Repo.get_by(App.HnswlibIndex, id: 1)
        |> case do
          nil ->
            {:stop, {:error, "Incoherence on table"}}
```

I've made some changes to the code so the GenServer is not executed right away when we run `mix test`. Instead, we properly seed the database (inside `test/test_helper.exs`) and then start the `App.KnnIndex` GenServer. 

This got it to work (the app still works on production/dev and does not yield "Incoherence on table" error it did before). However, the test fail. The following error is logged:

```
04:53:28.320 request_id=F7JnUXdn4cwgF7IAAAEC [warning] ❌ Index Integrity Error
```

I assume it's because I'm not properly seeding the data, so there must be an incongruency in the test data that is being seeded. 

Do you have any idea on how to surpass this? You can see by my last commit [`5bff3c9` (#47)](https://github.com/dwyl/image-classifier/pull/47/commits/5bff3c9ee0dd10ad2917a058c99e5f8463002409) to see how I'm seeding the data. I've simply used `DBeaver` to see the data and copied. I know this may not work straight away but do you know any way of adding the index to the test database so the tests work? 🤔

ndrean commented:
The reason is the missing "indexes.bin" file or using a bad one.

The "Integrity test" checks if the length of the `App.Image` table is the same as the current count of the HNSWLlib Index file.

I removed the `on_mount` and integrated it in the GenServer `init`. More linear.

Recall that the source of truth is the DB.
Since I don't know the state of the Fly.Io Filesystem, you check for coherence between the DB and the "indexes.bin" file (existing or not) from the Filesystem.

For tests: 
- you should load a given image,
- and copy the resulting HNSWLib Index file - say priv/static/indexes_test.bin" 

Then:
- if you `File.rm("indexes.bin")` and start the GS. IT should copy from the DB into the FS

- if you `File.copy("indexes_test.bin", "indexes.bin")` and start the GS, it should be fine.

LuchoTurtle commented:
Hmm 🤔 , ok. So, I have tried seeding only the `index` table and it gets an `Incoherence on table` error. Understandable.

But even if I have both of the tables empty, and `priv/static/uploads/indexes.bin` **removed** - basically a clean slate, I get a `"Index seems to be corrupted or unsupported"` error 🤔

ndrean commented:
Which file are you inserting in the `%HnswlibIndex{}`? It must be an index file generated by HNSWLib.

LuchoTurtle commented:
Thank you for the code. I tried it but subsequent runs with `mix test` made it crash. I've made some adjustments so the test env and prod/dev are separate and we can run those separately without conflicting `indexes.bin` files. 

Although one test is failing, now they always run. Every time you run `mix test`, all the test database is cleared and `indexes_test.bin` is forced to be created. After the tests are run, this index file is deleted. 

However, I do notice in the `Image` table that after the tests run (where one image is uploaded to be predicted), instead of only one row being created, *two are*: 

<img width="1316" alt="image" src="https://github.com/dwyl/image-classifier/assets/17494745/9b0b8a9d-d07f-4ec0-a0c9-79915811acd2">

I saw that this was because in `page_live.ex#195`, you're saving a partial image, which creates a "dummy" image with null values.

```elixir
        # save partial image
        App.Image.insert(image_info)
        |> case do
          {:ok, _} ->
            image_info =
              Map.merge(image_info, %{
                file_binary: file_binary
              })

            {:ok, %{tensor: tensor, image_info: image_info, path: path}}

          {:error, changeset} ->
            dbg(changeset.errors)
            {:error, changeset.errors}
        end
        |> handle_upload()
```

Why do we do this? Why is this needed (it wasn't before). When we do this, since the `KnnIndex` genserver checks for the length of the the `Images` table and the `Index` file, won't this create issues with data integrity?

ndrean commented:
Yes, you are right, I forgot to delete the "indexes.bin" when I commited .

I do a first partial save to "lock the sha" so that no one can upload the same picture since you "check_sha1" at line 178 before saving.
The second save updates it with the URL, the caption and the HNSWLib entry order.

The GenServer checks for the length only when it starts. I believe it starts only once during the tests.

ndrean commented:
The audio test will fail because it looks for the wrong div. Add an id:

"page_live.html.heex"
```elixir
<div
        class="flex mt-2 space-x-1.5 items-center font-bold text-gray-900 text-xl"
        id="audio-test"
                 ^^^
      >
        <span>Transcription: </span>
        <AppWeb.Spinner.spin spin={@speech_spin} />
        <%= if @transcription do %>
          <span id="output" class="text-gray-700 font-light"><%= @transcription %></span>
        <% else %>
          <span class="text-gray-300 font-light">Waiting for audio input.</span>
        <% end %>
      </div>
```

and use it:

```elixir
test "uploading an audio file", %{conn: conn} do
    {:ok, lv, html} = live(conn, ~p"/")
    assert html =~ "Caption your image!"

    # Get file and add it to the form
    file =
      [:code.priv_dir(:app), "static", "audio", "itwillallbeok.mp3"]
      |> Path.join()
      |> build_upload("audio/mp3")

    audio = file_input(lv, "#audio-upload-form", :speech, [file])

    # The transcription should be empty
    # assert render(lv) |> Floki.find("#audio-test") |> Floki.text() ==
    #          "Transcription: Waiting for audio input."

    # Should show an uploaded local file
    assert render_upload(audio, file.name)

    # Wait for the audio prediction to end
    AppWeb.SupervisorSupport.wait_for_completion()

    # A prediction should have occurred and the label should be shown with the audio transcription
    assert render(lv) |> Floki.find("#audio-test") |> Floki.text() =~
             "Sometimes, the inner voice is encouraging"
  end
```

The tests sometimes works, sometimes not.

ndrean commented:
The `with` statement makes the happy path reasonably easy to follow, with ckear functions, but at the expense of having a labyrinth of edge cases.
Not so sure anymore if its worth it.

ndrean commented:
Just a side note. I don't know why I wrote this comment above on adding an id for the audio file test 🤔🤯 I already did put the id "output" on the `<span>`.  Got confused...

ndrean commented:
@LuchoTurtle 
I added most of the relevant tests for the GenServer.
A few edge cases missing. 85%

LuchoTurtle commented:
I've merged your changes with local changes and I'm getting `76.2%` instead :P

ndrean commented:
@LuchoTurtle 
Yes, sorry
Can eliminate edge cases  from Coveralls such as those related to `File.write` ?

LuchoTurtle commented:
> Can eliminate edge cases from Coveralls such as those related to `File.write` ?
Can you be more specific? You mean ignore lines from edge cases?

ndrean commented:
```elixir
#App.KnnIndex.ex

def init(args) do
[...]
case File.exists?(index_path) do
 false ->
App.HnswlibIndex.maybe_load_index_from_db(space, @dim, @max_elements)
|> case do
  {:ok, index, index_schema} -> {:ok, {index, index_schema, space}}
  {:error, msg} -> {:stop, {:error, msg}}
#  ^^^ I am unable to reach this one with tests so I want to eliminate this case
end

ndrean commented:
I pulled

LuchoTurtle commented:
Awesome :) 

We're super close. After the tests are done, I'll just check the readme and basic formatting and we're done :).

Thanks for the effort put in so far ❤️

ndrean commented:
Thanks for your patience. It was a pleasure as I learn testing with you. Almost 2 years of Elixir for me.

Note: you notice I introduced args in the GS to be able to run tests.

ndrean commented:
@LuchoTurtle 

I can't get the following test running: the code supposed to stop the Audio transcription from running if no image is loaded or available.
It is related to this part of "page_live.ex":

```elixir
def handle_info({ref, %{chunks: [%{text: text}]} = _result}, %{assigns: assigns} = socket)
      when assigns.audio_ref == ref do
    Process.demonitor(ref, [:flush])
    File.rm!(assigns.tmp_wave)

    # compute an normed embedding (cosine case only) on the text result
    # and returns an App.Image{} as the result of a "knn_search"
    with {:not_empty_index, :ok} <-
             ^^^
           {:not_empty_index, App.KnnIndex.not_empty_index() |> dbg()},
         %{embedding: input_embedding} <-
           Nx.Serving.batched_run(Embedding, text),
         %Nx.Tensor{} = normed_input_embedding <-
           Nx.divide(input_embedding, Nx.LinAlg.norm(input_embedding)),
         %App.Image{} = result <-
           App.KnnIndex.knn_search(normed_input_embedding) do
      {:noreply,
       assign(socket,
         transcription: String.trim(text),
         micro_off: false,
         speech_spin: false,
         search_result: result,
         audio_ref: nil,
         tmp_wave: @tmp_wav
       )}
    else
      # record without entries
      {:not_empty_index, :error} ->
          ^^^
        Logger.warning("No entries in Index")

        {:noreply,
         socket
         |> push_event("toast", %{message: "No images yet"})
         |> assign(
           micro_off: false,
           search_result: nil,
           speech_spin: false,
           audio_ref: nil,
           tmp_wave: @tmp_wav
         )}
        [....]
```

To test this, I have(?)to use a fresh new app and load an empty index file. 
This empty index  is generated by the snippet below (from App.HnswlibIndex.ex, not pushed).

```elixir
if Mix.env() == :test do
      empty_index =
        Application.app_dir(:app, ["priv", "static", "uploads"])
        |> Path.join("indexes_empty.bin")

      HNSWLib.Index.save_index(index, empty_index)
    end
```

The test further below runs ok but does not trigger the case "not_empty" in the GenServer:

```elixir
def handle_call(:not_empty, _, {index, _, _} = state) do
    case HNSWLib.Index.get_current_count(index) |> dbg() do
      {:ok, 0} ->
        Logger.warning("Empty index")
        ^^^ fails!
        {:reply, :error, state}

      {:ok, _} ->
        ^^^ uses this switch
        {:reply, :ok, state}
    end
  end
```

The test is:

```elixir
test "do not run audio if image bank is empty", %{conn: conn} do
    path = set_path("indexes_empty.bin")
    App.Application.start(:normal, [])
    Supervisor.which_children(App.Supervisor)
    App.Repo.delete_all(App.HnswlibIndex)
    App.Repo.delete_all(App.Image)

    %App.HnswlibIndex{}
    |> App.HnswlibIndex.changeset(%{
      lock_version: 2,
      file: nil,
      id: 1
    })
    |> App.Repo.insert()

    {:ok, _} = 
       Supervisor.start_child(
      App.Supervisor,
      {App.KnnIndex, [space: :cosine, index: path]}
    )

    {:ok, lv, _html} = live(conn, ~p"/")

    # Get audio file and add it to the form
    [:code.priv_dir(:app), "static", "audio", "itwillallbeok.mp3"]
    |> Path.join()
    |> build_upload("audio/mp3")

    #   # Wait for the audio prediction to end
    AppWeb.SupervisorSupport.wait_for_completion()

    # A prediction should have occurred and the label should be shown with the audio transcription
    assert render(lv) |> Floki.find("#output") |> Floki.text() =~
             ""

   # I should have a warning log "Empty Index"
  end
```

I suspect it does not work because I use a `conn` from the param that comes probably(?) from `use AppWeb.ConnCase`.
So I wondered if Is there a way to build a conn and not use it ex_nihilo as a param?
I hope I am clear.

LuchoTurtle commented:
If the index already has two images, it's because the other tests are the culprit, which makes sense. The tests are run sequentially, so the DB is bound to have data, as well as the index file.

But I have a sneaking suspicion that the reason you have 2 images, it's because of the example images that are shown after a period of inactivity, and it's being translated in that new test you've created.

ndrean commented:
@LuchoTurtle 
I pushed the code. The "indexes_empty.bin" should be created when running `mix test` but I commited it.

LuchoTurtle commented:
Just pushed the fix. The reason the test was failing was two-fold:

- because `AppWeb.SupervisorSupport.wait_for_completion()` waits for all the `pids` to finish, I'm adding two `render_async` so the test has time to spawn the `audio transcription` and `embedding` async tasks.
- you forgot `render_upload` to basically upload the audio.

It should work now :)

ndrean commented:
@LuchoTurtle 
I am at a point where I removed or omitted switch cases to reach 100% coverage. What do you think?

LuchoTurtle commented:
If they're unreachable with tests, it means that the edge cases are probably useless. But if you feel like there's any edge case that may be frequent, we may need to test it, even if it's thru mocking or just directly unit testing it.

Feel free to remove the edge cases that you may find "extra" :)

ndrean commented:
@LuchoTurtle 
I reached a limit of 97%. Can't reach a failure on insertions to the db (eg Ecto.Multi) and the S3 upload.
I am sure I experienced these cases but can't reproduce them.
I suggest you take a look at the coveralls.html.

LuchoTurtle commented:
I've fixed some tests, some edge cases and brought the coverage to 100%.

I've removed the commented code that was unused.

Is the README ready to be reviewed? 👀

LuchoTurtle commented:
@ndrean I've gone through the `README` and really thought it was insightful! 

I've formatted and numbered the chapters so it follows the same structure as before. 

I see that you left it midway. Do you want me to complete the README or do you have any local changes that you haven't pushed or wish to push? 

Let me know so we can finally (at long last :P) get this PR merged :D

ndrean commented:
@LuchoTurtle 
I am without a computer for a few days. 
I did not pursue the Readme because:
- the code of the  first part maybe  be impacted by the code modifications of the second part. In order not to refactor to heavily in the 2d part. I thought it was better to integrate changed directly but again only if you approve the code
- and the same applies for the second part, except that I added some motivations on the choices, eg why Hnswlib and not pgvector with ANN  , or why a Genserver to handle the Index. 
- I wanted the tests to be done


Feel free to pursue it.

I also wanted to see if loading each model via a Genserver with a handle_continue would speed up the cold boot  process. I will test this and let you know.

LuchoTurtle commented:
There's a reason for this ☺️

Here's a link for a small explanation on this:
https://emmanuelbernard.com/blog/2013/08/08/one-line-per-idea/

Other than that, I think this PR is ready to be merged, don't you agree?

ndrean commented:
Some remarks:

1) `Bandit`. I saw that Phoenix uses it by default, so I used it. However, it maybe not be a good idea due to [memory usage](https://elixirforum.com/t/how-to-troubleshoot-increased-memory-usage-in-bandit/61825)

2) ELXA: 
I updated `EXLA` and `hnswlib` and `nx` and I could make it work due to the error below

```bash
[warning] The on_load function for module Elixir.EXLA.NIF returned:
{:error, {:load_failed,
  ~c"Failed to load NIF library: 'dlopen(/Users/nevendrean/code/elixir/image-classifier/_build/dev/lib/exla/priv/libexla.so
```

I erased  "_build" and the local the cache, and tried to `mix compile` and it still did not work.
The Dockerfile does nothing special in the sense that it does not set env vars. 
However, I could not get it working until I set the env var `XLA_CACHE_DIR` as below (pointing to my cache).

```bash
 export XLA_CACHE_DIR=/Users/nevendrean/Library/Caches/xla/exla
```
I don't understand what happens.

3) Failling async model loading
I transformed the models loading into a generic Genserver that takes the model as an argument. It was supposed to off-load the processes loading in a `handle_continue`. These Genservers are set as childs to the Supervisor (use a map `%{id: :audio, start: {....}`).  It still see a sequential loading so it may not be the way to go.

LuchoTurtle commented:
@ndrean Thank you again for the awesome insight (as always!).

1 - I've made this change, thank you for the warning. Reverted.
2 - It's weird that you got this happening to you 🤔 . Regardless, although upgrading the dependencies is always appreciated, it seems that with the [new version of `Bumblebee`](https://github.com/elixir-nx/bumblebee/blob/main/CHANGELOG.md), some breaking changes have been introduced that may conflict with the work we've done here. 
3 - It's a nice idea to try and load the models async. But I feel like this PR is already massive and adds a ton of value. I don't think it's productive to delve into optimization in this PR, it's already bloated with many changes (great ones!), it's best to merge it and deal with each optimization separately. Again, we want to make this PR beginner-friendly, it's a tutorial meant to guide people :) 

Thank you for the awesome work and the **great value** delivered on this PR!

We'll make some changes on the HTML/CSS on a different view to make it more pleasing to the eye :) 

Thanks a lot!

ndrean commented:
@LuchoTurtle 
Note that I did test the async loading to try to (partly) improve the loading time of the models every time the Fly machine is pruned.

ndrean commented:
@LuchoTurtle 
If you want to use tabs, such as one with a file input, and one with audio input, I saved [some notes](https://dev.to/ndrean/breadcumbs-with-phoenix-liveview-2d40) on how one can do this by creating 2 components and some links to navigate between tabs. It is a simply `JS.hide` / `JS.show`.
